scripts/collate.py

```diff
@@ -42,52 +45,51 @@
         df = dbcan
 
     # kofam annotations
-    if filename.endswith('_kofam.tsv'):
+    elif filename.endswith('_kofam.tsv'):
         kofam = pd.read_csv(filename, sep="\t", index_col=False)
         # Prefix all columns with 'kofam.'
         kofam = kofam.rename(columns={"kofam_KO"      : "kofam.KEGG_KO",
                                       "kofam_Module"  : "kofam.KEGG_Module",
                                       "kofam_Pathway" : "kofam.KEGG_Pathway"})
         kofam.set_index("ID", inplace=True)
         df = kofam
+    else:
+        invalid_inputs.append(filename)
+        print("WARNING:", filename, "not supported", file=sys.stderr)
 
-# eggnog and kofam
-if (not eggnog.empty and not kofam.empty):
-    KOdf = pd.merge(eggnog,kofam, on="ID", how="outer").fillna("-")
-    eggnoglist = KOdf['eggNOG.KEGG_KO'].tolist()
-    kofamlist = KOdf['kofam.KEGG_KO'].tolist()
-    newlist = []
-    final = []
-    for i in range(len(kofamlist)):
-        mydict = {}
-        myline = eggnoglist[i]+","+kofamlist[i]
-        mylist = myline.split(",")
-        for i in mylist:
-            mydict[i] = ""
-        newlist = list(mydict.keys())
-        mystring = ""
-        if len(newlist) > 1:
-            if "-" in newlist:
-                newlist.remove("-")
-            mystring = ",".join(newlist)
-            final.append(mystring)
-        else:
-            mystring = "".join(newlist)
-            final.append(mystring)
-    KOdf["merged.KEGG_KO"] = final
-    df = KOdf
-# eggnog and dbcan, no kofam
-elif (not eggnog.empty and not dbcan.empty and kofam.empty):
-    del df
-    df = pd.merge(eggnog,dbcan, on="ID", how="outer").fillna("-")
-# kofam and dbcan, no eggnog
-elif (not kofam.empty and not dbcan.empty and eggnog.empty):
-    del df
-    df = pd.merge(dbcan,kofam, on="ID", how="outer").fillna("-")
+for fn in invalid_inputs:
+    input_annotations.remove(fn)
 
-# eggnog, kofam and dbcan
-if (not eggnog.empty and not kofam.empty and not dbcan.empty):
-    del df
-    df = pd.merge(KOdf,dbcan, on="ID", how="outer").fillna("-")
-
-df.to_csv(sys.stdout, sep="\t")
+if len(input_annotations) > 1:
+    # eggnog and kofam
+    if (not eggnog.empty and not kofam.empty):
+        KOdf = pd.merge(eggnog,kofam, on="ID", how="outer").fillna("-")
+        eggnoglist = KOdf['eggNOG.KEGG_KO'].tolist()
+        kofamlist = KOdf['kofam.KEGG_KO'].tolist()
+        final_kos = []
+        for i in range(len(kofamlist)):
+            mergedlist = eggnoglist[i].split(",") + kofamlist[i].split(",")
+            uniqmergedlist = list(set(mergedlist))
+            mystring = ""
+            if len(uniqmergedlist) > 1:
+                if "-" in uniqmergedlist:
+                    uniqmergedlist.remove("-")
+                mystring = ",".join(uniqmergedlist)
+            else:
+                mystring = uniqmergedlist[0]
+            final_kos.append(mystring)
```

Comment:
Will the order be same between different instances? Could there be randomness involved? We need the order to be preserved when we perform benchmarks and check for identical outputs.

Reply:
It's a set() so order is not guaranteed. I've seen suggestions that setting the seed in PYTHONHASHSEED would make the underlying hash-table reproducible for the same implementation and the same input. I can try that, or re-write to do membership testing on eggnog vs kofam. Which would you prefer?

Reply:
How about adding a `sort()` in front instead, so that the order is consistent?

Reply:
Too late, I did the list version. It's now outputing the same tsv as the previous collate.py